python/cudf/cudf/core/multiindex.py

```diff
@@ -392,13 +393,124 @@ def _popn(self, n):
+            current_row = row_tuples[row_idx]
+            current_row_repr = line.split(", ")
+
+            # If there is mismatch in lengths, that means there
+            # is a string in multi-Index which contains ", "
+            # Due to this reason we will have to correctly
+            # calculate the splits by inspecting values in current_row
+            if len(current_row_repr) != len(current_row):
+                correct_row_repr = []
+                repr_idx = 0
+                for val in current_row:
+                    if not isinstance(val, str) or (
+                        isinstance(val, str) and ", " not in val
+                    ):
+                        correct_row_repr.append(current_row_repr[repr_idx])
+                        repr_idx += 1
+                    else:
+                        split_count = val.count(", ")
+                        correct_row_repr.append(
+                            ", ".join(
+                                current_row_repr[
+                                    repr_idx : repr_idx + split_count + 1
+                                ]
+                            )
+                        )
+                        repr_idx += split_count + 1
+                current_row_repr = correct_row_repr
+
+            for col_idx, value in enumerate(current_row):
+                if value is pd.NA:
+                    if col_idx in date_time_col_indices:
+                        current_row_repr[col_idx] = current_row_repr[
+                            col_idx
+                        ].replace("NaT", cudf._NA_REP)
+                    else:
+                        current_row_repr[col_idx] = current_row_repr[
+                            col_idx
+                        ].replace("nan", cudf._NA_REP)
+                elif value is np.nan:
+                    # Leave nan's as is.
+                    pass
+                elif cols_nulls[col_idx]:
+                    # Need to pad the values with extra space
+                    # as pandas MultiIndex return 3 character
+                    # string "nan", which are replaced by
+                    # 4 character string "<NA>", hence the
+                    # adjustment is needed for aligned output.
+                    if col_idx == 0:
+                        current_row_repr[col_idx] = current_row_repr[
+                            col_idx
+                        ].replace("(", "( ", 1)
+                    elif col_idx == len(current_row) - 1:
+                        current_row_repr[col_idx] = (
+                            " " + current_row_repr[col_idx]
+                        )
+                    else:
+                        current_row_repr[col_idx] = (
+                            " " + current_row_repr[col_idx]
+                        )
+
+            new_lines.append(", ".join(current_row_repr))
+        new_lines.append(lines[-1])
+        lines = new_lines
+        return lines
```

Comment:
This seems quite inefficient. We're converting the entire source dataframe to string dtype, and copying to pandas instead of only the necessary rows. Additionally we're iterating over all of the rows of the MultiIndex.

I think we should be able to take a more efficient approach here and figure out where there's nulls in `preprocess` and then modify the underlying array of the Pandas MultiIndex to fix the repr.

Reply:
Replaced this logic to go via pyarrow since pyarrow nicely supports having `None` & `np.nan` in their float columns.